mobile-browser-based-version/src/helpers/training/training_manager.js

```diff
@@ -144,79 +148,215 @@ export class TrainingManager {
   async _trainLocally() {
     const info = this.task.trainingInformation;
 
-    await this.model.fit(this.data, this.labels, {
-      initialEpoch: this.model.getUserDefinedMetadata().epoch,
-      epochs: info.epochs ?? MANY_EPOCHS,
-      batchSize: info.batchSize,
-      validationSplit: info.validationSplit,
-      shuffle: true,
-      callbacks: {
-        onEpochEnd: async (epoch, logs) => {
-          this.trainingInformant.updateGraph(
-            epoch + 1,
-            (logs.val_acc * 100).toFixed(2),
-            (logs.acc * 100).toFixed(2)
-          );
-          console.log(
-            `EPOCH (${epoch + 1}):
+    if (info.batchwise) {
+      console.log(
+        'Memory efficient training mode is used, data preprocessing is executed batch wise'
+      );
+      // Creation of Dataset objects for training
+      const trainData = tf.data
+        .generator(
+          datasetGenerator(
+            this.data,
+            this.labels,
+            0,
+            this.data.shape[0] * (1 - info.validationSplit),
+            info
+          )
+        )
+        .batch(info.batchSize);
+
+      const valData = tf.data
+        .generator(
+          datasetGenerator(
+            this.data,
+            this.labels,
+            Math.floor(this.data.shape[0] * (1 - info.validationSplit)),
+            this.data.shape[0],
+            info
+          )
+        )
+        .batch(info.batchSize);
+      await this.model.fitDataset(trainData, {
+        epochs: info.epochs,
+        validationData: valData,
+        callbacks: {
+          onEpochEnd: async (epoch, logs) => {
+            this.trainingInformant.updateGraph(
+              epoch + 1,
+              (logs.val_acc * 100).toFixed(2),
+              (logs.acc * 100).toFixed(2)
+            );
+            console.log(
+              `EPOCH (${epoch + 1}):
+            Train Accuracy: ${(logs.acc * 100).toFixed(2)},
+            Val Accuracy:  ${(logs.val_acc * 100).toFixed(2)}\n`
+            );
+            if (this.useIndexedDB) {
+              this.model.setUserDefinedMetadata({ epoch: epoch + 1 });
+              await updateWorkingModel(
+                this.task.taskID,
+                info.modelID,
+                this.model
+              );
+            }
+            if (this.stopTrainingRequested) {
+              this.model.stopTraining = true;
+              this.stopTrainingRequested = false;
+            }
+          },
+        },
+      });
+    } else {
+      console.log(
+        'Fast training mode is used, data preprocessing is executed on the entire dataset at once'
+      );
+
+      const tensor = preprocessData(this.data, info);
+
+      await this.model.fit(tensor, this.labels, {
```

Comment:
There is an if else here, where each does model fitting in a different way. They are each quite long, could you make a function for each:

if (info.batchwise) {
  modelFitDataBatchWise(...)
} else {
  modelFitData(....)
}

Something along these lines (maybe you can find better names).